oxidation/libparsec/crates/crypto/src/common/secret.rs

```diff
@@ -1,72 +1,92 @@
 // Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
 
-use base32::Alphabet;
-use serde::Serialize;
-use serde_bytes::Bytes;
-use std::fmt::Debug;
+/// 34 symbols to 32 values due to 0/O and 1/I
+pub(crate) const RECOVERY_PASSPHRASE_SYMBOLS: [char; 34] = [
+    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S',
+    'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '0', '1', '2', '3', '4', '5', '6', '7',
+];
 
-use crate::{CryptoError, RECOVERY_PASSPHRASE_SYMBOLS};
+macro_rules! impl_secret_key {
+    ($key: ident) => {
+        crate::impl_key_debug!($key);
 
-pub trait SecretKeyTrait:
-    AsRef<[u8]>
-    + for<'a> TryFrom<&'a [u8], Error = CryptoError>
-    + for<'a> TryFrom<&'a Bytes, Error = CryptoError>
-    + From<[u8; 32]>
-    + Serialize
-    + Debug
-{
-    const ALGORITHM: &'static str = "xsalsa20poly1305";
-    const SIZE: usize = 32;
+        impl $key {
+            pub fn generate_recovery_passphrase() -> (String, Self) {
+                let key = Self::generate();
+                let b32 =
+                    ::base32::encode(::base32::Alphabet::RFC4648 { padding: false }, key.as_ref());
+                let passphrase = b32
+                    .as_bytes()
+                    .chunks(4)
+                    .map(std::str::from_utf8)
+                    .collect::<Result<Vec<&str>, _>>()
+                    .expect("Unreachable because variable b32 is a utf8 string")
+                    .join("-");
+                (passphrase, key)
+            }
 
-    fn generate() -> Self;
-    fn encrypt(&self, data: &[u8]) -> Vec<u8>;
-    fn decrypt(&self, ciphered: &[u8]) -> Result<Vec<u8>, CryptoError>;
-    fn hmac(&self, data: &[u8], digest_size: usize) -> Vec<u8>;
+            pub fn from_recovery_passphrase(passphrase: &str) -> crate::CryptoResult<Self> {
+                // Lowercase is not allowed in theory, but it's too tempting to fix this here ;-)
+                let passphrase = passphrase.to_uppercase();
+                // Filter out any unknown characters, this is typically useful to remove
+                // the `-` and whitespaces.
+                // Note we also discard possible typos from the user (for instance if he types
+                // a `8` or a `9`), but this is no big deal given 1) it should not happen
+                // because GUI should use `RECOVERY_PASSPHRASE_SYMBOLS` to prevent user
+                // from being able to provide invalid characters and 2) it will most likely
+                // lead to a bad password anyway
+                let mut b32 = passphrase
+                    .chars()
+                    .filter(|c| crate::common::RECOVERY_PASSPHRASE_SYMBOLS.contains(c))
```

Comment:
Isn't is better to use a closure like this instead of iterating in an array?
```rust
|c| c >= 'A' && c <= 'Z' || c >= '0' && c <= '9'
```

Reply:
match is better

Reply:
```suggestion
    match c {
        'A' ..= 'Z' | '0' ..= '9' => true,
        _ => false
    }
```

Reply:
Done

Reply:
Note that '8' and '9' are not in `RECOVERY_PASSPHRASE_SYMBOLS`

Reply:
Oh mb, I didn't noticed it